client/client.go

```diff
@@ -2419,6 +2419,24 @@ OUTER:
 				}
 			}
 
+			// It is possible that Alloc.GetAllocs hits a different server than
+			// Node.GetClientAllocs which returns older results.
+			if allocsResp.Index <= allocsReq.MinQueryIndex {
```

Comment:
Shouldn't this be `<`? Unlike the case above with `Node.GetClientAllocs`, we want to make sure the index is at least the _same_ index as the previous `Node.GetClientAllocs` RPC and we've set the `MinQueryIndex` to `pullIndex - 1` to ensure that. That is, the following is expected behavior if nothing has changed between the two calls:

* `Node.GetClientAllocs` returns index 10.
* we set `MinQueryIndex` to index 9, because we don't want to wait for index 11.
* `Alloc.GetAllocs` returns index 10.

This all springs from the unfortunate name of `MinQueryIndex`, which really is "this is the last index I saw":

```go
 	// If set, wait until query exceeds given index. Must be provided
	// with MaxQueryTime.
	MinQueryIndex uint64
```

Reply:
Nice catch. I think I was fooled by the fact that we're inconsistent (line `2480` in turn does not add the `-1` to the index which makes the former check `<=` instead of `<`). Should we pick one, stick to it and be consistent throughout?

Reply:
Ideally it'd have been one query that gets us everything we need, but this reduces the total amount of data we have to send/receive by quite a bit for clients with a large amount of allocations.

Reply:
Yup, you're right. Sorry about that, in retrospect It was a silly idea on my part. Thanks for the explanation!

Reply:
Great catch. I'll update and make an attempt at a comment.

Reply:
Actually due to the -1 shouldn't this be a `<=`? In Tim's example:

1. Node.GetClientAllocs -> 10
2. 9 -> Alloc.GetAllocs
3. Alloc.GetAllocs -> 10

`<=` is fine here because 10 > 9 and equal to the index you got before. This is the happy (and hopefully common) path.

But let's say you get unlucky and hit time travel:

1. Node.GetClientAllocs -> 10
2. 9 -> Alloc.GetAllocs
3. Alloc.GetAllocs -> 9

Only retrying on `<` here would allow the client to use allocs at index 9 which might differ from the state the previous RPC saw at index 10!

Therefore I think due to the -1 we can maintain the `<=` comparison because it means "at least at the index the previous rpc saw."

Reply:
Ah, you're right, we're comparing the response to the `MinQueryIndex`, not the `pullIndex`.